refer *load-truename* first to avoid implementation specific behavior.

takagi opened the pull request:
**Situation**
Loading split-sequence within top level `eval-when` form causes an error on SBCL.

    $ cat foo.lisp
    (eval-when (:compile-toplevel :load-toplevel :execute)
      (asdf:oos 'asdf:load-op :split-sequence))

    * (compile-file "foo.lisp")    ; READ error on SBCL

(Although you may look AVER failed, it is not essential for this time and already reported on SBCL's bug database. Note just an READ error.)

I want to do this to use Common Lisp codes as scripts called from UNIX shell using [roswell](https://github.com/snmsts/roswell).

This problem does not happen on CCL.

**Cause**
The cause seems that the exact value of `*compile-file-pathname*` is implementation dependent. I checked it with the following code.

    $ cat split-sequence.asd
    ...
      :version #.(progn
                   (print *compile-file-pathname*)
                   (print *load-truename*)
                   (with-open-file (f (merge-pathnames "version.lisp-expr"
                                                       (or *compile-file-pathname*
                                                           *load-truename*)))
                     (read f)))
    ...

On SBCL, `*compile-file-pathname*` has full path for the file being compiled. It is merged with "version.lisp-expr" to return /User/mtakagi/Desktop/version.lisp-expr, causing file not found.

    * (compile-file "foo.lisp")
    #P"/Users/mtakagi/Desktop/foo.lisp"
    #P"/Users/mtakagi/Developer/quicklisp/local-projects/split-sequence/split-sequence.asd"
    ...

On CCL, `*compile-file-pathname*` has just name for the file. It is merged with "version.lisp-expr" to return split-sequence/version.lisp-expr, expected path.

    ? (compile-file "foo.lisp")
    #P"foo.lisp"
    #P"/Users/mtakagi/Developer/quicklisp/local-projects/split-sequence/split-sequence.asd"
    ...

**Solution**
Refering `*load-pathname*` first will solve this problem. As far as I tested this change on various cases on SBCL and CCL, it worked well.

    $ cat split-sequence.asd
    ...
      :version #.(with-open-file (f (merge-pathnames "version.lisp-expr"
                                                     (or *load-truename*
                                                         *compile-file-pathname*)))
                     (read f))
    ...

**Failed ideas**
Using `*default-pathname-defaults*` does not work because it does not contain expected path when compile-file split-sequence.asd within `eval-when` form.

    $ cat split-sequence.asd
    ...
      :version #.(progn
                 (print *default-pathname-defaults*)
                 (with-open-file (f (merge-pathnames "version.lisp-expr"
                                                     *default-pathname-defaults*))
                     (read f)))
    ...

    $ cat bar.lisp
    (eval-when (:compile-toplevel :load-toplevel :execute)
      (compile-file "/path/to/split-sequence/split-sequence.asd"))

    * (compile-file "bar.lisp")
    #P"/Users/mtakagi/Desktop/"
    ; error, version.lisp-expr not found

As another, using `asdf:system-source-directory` causes infinite loop.

luismbo commented:
Are asd files ever compiled? I.e., should we be looking at `*compile-file-truename* at all?

sionescu commented:
That was a hack for before :version-file was added to ASDF. Now that all 
major Lisps have upgraded to an ASDF that supports :version-file, we should 
use that. See example in IOlib. I'm in vacation so can't do it right now.

--
Stelian Ionescu a.k.a. fe[nl]ix
Quidquid latine dictum sit, altum videtur



On June 23, 2015 10:27:56 AM Luís Oliveira <notifications@github.com> wrote:

> Are asd files ever compiled? I.e., should we be looking at 
> `*compile-file-truename* at all?
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/sharplispers/split-sequence/pull/4#issuecomment-114405367

takagi commented:
@luismbo 
I think, as you say, `*compile-file-pathname*` is not needed because ASDF system file should be always `asdf:load-op`ed. Actually `*load-truename*` is nil in the following case, it makes no sense.

    (compile-file "split-sequence/split-sequence.asd")

I tested the following cases on SBCL and CCL, everything works well.
- `(load "foo.lisp")`
- `(progn (compile-file "foo.lisp") (load "foo.fasl"))`
- `(load "bar.lisp")`
- `(compile-file "bar.lisp")`

where

    $ cat foo.lisp
    (asdf:oos 'asdf:load-op :split-sequence)

    $ cat bar.lisp
    (eval-when (:compile-toplevel :load-toplevel :execute)
      (asdf:oos 'asdf:load-op :split-sequence))

takagi commented:
@sionescu Cool. Then I will do it seeing IOlib so please review it.

takagi commented:
I fixed to use :version-file.
@sionescu Is this fine?
@luismbo Would you merge it after @sionescu confirmed?

luismbo commented:
Cherry picked your last commit and pushed da0b5b0, thanks!
